File: 2025/12/04/cicd-agentic-ai/typescript/src/agent.ts
/** Agent implementation with versioning and tool management */

export enum AgentRole {
  PLANNER = "planner",
  WORKER = "worker",
  CRITIC = "critic",
  ROUTER = "router"
}

export interface ModelConfig {
  model: string;
  temperature: number;
  maxTokens?: number;
}

export interface AgentResult {
  success: boolean;
  result?: any;
  tools_called: string[];
  latency_ms: number;
  agent_version: string;
  error?: string;
}

export class Agent {
  role: AgentRole;
  modelConfig: ModelConfig;
  tools: string[];
  version: string;
  toolRegistry: Map<string, Function>;
  toolsCalled: string[];

  constructor(
    role: AgentRole,
    modelConfig: ModelConfig,
    tools: string[],
    version: string,
    toolRegistry?: Map<string, Function>
  ) {
    this.role = role;
    this.modelConfig = modelConfig;
    this.tools = tools;
    this.version = version;
    this.toolRegistry = toolRegistry || new Map();
    this.toolsCalled = [];
  }

  async run(inputData: Record<string, any>): Promise<AgentResult> {
    const startTime = Date.now();

    try {
      const result = await this.execute(inputData);
      const latencyMs = Date.now() - startTime;

      return {
        success: true,
        result,
        tools_called: this.toolsCalled,
        latency_ms: latencyMs,
        agent_version: this.version
      };
    } catch (error: any) {
      const latencyMs = Date.now() - startTime;
      return {
        success: false,
        error: error.message,
        tools_called: this.toolsCalled,
        latency_ms: latencyMs,
        agent_version: this.version
      };
    }
  }

  private async execute(inputData: Record<string, any>): Promise<any> {
    switch (this.role) {
      case AgentRole.PLANNER:
        return this.plan(inputData);
      case AgentRole.WORKER:
        return this.work(inputData);
      case AgentRole.CRITIC:
        return this.critique(inputData);
      default:
        return { status: "unknown_role" };
    }
  }

  private plan(inputData: Record<string, any>): any {
    const plan: any = {
      steps: [],
      tools_needed: []
    };

    const input = inputData.input || "";
    if (input.toLowerCase().includes("reset password")) {
      plan.steps = ["search_kb", "create_ticket"];
      plan.tools_needed = ["search_kb", "create_ticket"];
    }

    for (const tool of plan.tools_needed) {
      if (this.tools.includes(tool)) {
        this.toolsCalled.push(tool);
      }
    }

    return { plan };
  }

  private async work(inputData: Record<string, any>): Promise<any> {
    const toolsToCall = inputData.tools || [];
    const results: Record<string, any> = {};

    for (const tool of toolsToCall) {
      if (this.tools.includes(tool) && this.toolRegistry.has(tool)) {
        try {
          const toolFn = this.toolRegistry.get(tool)!;
          const result = await toolFn(inputData);
          results[tool] = result;
          this.toolsCalled.push(tool);
        } catch (error: any) {
          results[tool] = { error: error.message };
        }
      }
    }

    return { results };
  }

  private critique(inputData: Record<string, any>): any {
    return {
      critique: "Work looks good",
      suggestions: []
    };
  }
}

File: 2025/12/04/cicd-agentic-ai/python/src/agent.py
"""Agent implementation with versioning and tool management"""
from typing import List, Dict, Any, Optional
from enum import Enum
import json
import time


class AgentRole(str, Enum):
    PLANNER = "planner"
    WORKER = "worker"
    CRITIC = "critic"
    ROUTER = "router"


class Agent:
    """Agent with explicit versioning and tool management"""
    
    def __init__(
        self,
        role: AgentRole,
        model_config: Dict[str, Any],
        tools: List[str],
        version: str,
        tool_registry: Optional[Dict[str, callable]] = None
    ):
        self.role = role
        self.model_config = model_config
        self.tools = tools
        self.version = version
        self.tool_registry = tool_registry or {}
        self.tools_called = []
    
    def run(self, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """Run agent with input data"""
        start_time = time.time()
        
        try:
            # Simulate agent execution
            result = self._execute(input_data)
            
            latency_ms = int((time.time() - start_time) * 1000)
            
            return {
                "success": True,
                "result": result,
                "tools_called": self.tools_called,
                "latency_ms": latency_ms,
                "agent_version": self.version
            }
        except Exception as e:
            latency_ms = int((time.time() - start_time) * 1000)
            return {
                "success": False,
                "error": str(e),
                "tools_called": self.tools_called,
                "latency_ms": latency_ms,
                "agent_version": self.version
            }
    
    def _execute(self, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """Execute agent logic"""
        # Simulate agent decision-making
        if self.role == AgentRole.PLANNER:
            return self._plan(input_data)
        elif self.role == AgentRole.WORKER:
            return self._work(input_data)
        elif self.role == AgentRole.CRITIC:
            return self._critique(input_data)
        else:
            return {"status": "unknown_role"}
    
    def _plan(self, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """Planner agent logic"""
        plan = {
            "steps": [],
            "tools_needed": []
        }
        
        # Simple planning logic
        if "reset password" in input_data.get("input", "").lower():
            plan["steps"] = ["search_kb", "create_ticket"]
            plan["tools_needed"] = ["search_kb", "create_ticket"]
        
        # Track tool calls
        for tool in plan["tools_needed"]:
            if tool in self.tools:
                self.tools_called.append(tool)
        
        return {"plan": plan}
    
    def _work(self, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """Worker agent logic"""
        tools_to_call = input_data.get("tools", [])
        results = {}
        
        for tool in tools_to_call:
            if tool in self.tools and tool in self.tool_registry:
                try:
                    result = self.tool_registry[tool](input_data)
                    results[tool] = result
                    self.tools_called.append(tool)
                except Exception as e:
                    results[tool] = {"error": str(e)}
        
        return {"results": results}
    
    def _critique(self, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """Critic agent logic"""
        return {
            "critique": "Work looks good",
            "suggestions": []
        }

